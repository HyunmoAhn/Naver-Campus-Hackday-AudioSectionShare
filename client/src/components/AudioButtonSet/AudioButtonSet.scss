.AudioButtonSet__play-btn {
  background-color: transparent;
  border: 2px solid #BABABA;
  border-radius: 50%;
  color: white;
  width: 35px;
  height: 35px;
  margin-top: 10px;
  font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
}

.AudioButtonSet__section-btn {
  position: relative;
  background-color: transparent;
  border: 0;
  color: white;
  left: -15px;

  .active {
    color: #66D36E;
  }
}

.AudioButtonSet__toggle-btn {
  position: relative;
  background-color: transparent;
  border: 0;
  color: white;
  left: 15px;

  .active {
    color: #66D36E;
  }
}

.AudioButtonSet__share-btn {
  position: relative;
  width: 100px;
  margin-left: -100px;
  background-color: transparent;
  border: 0;
  left: 150px;
  color: white;

  .active {
    color: #66D36E;
  }
}

.AudioButtonSet__section-loop-cancel {
  background-color: transparent;
  border: 1px solid white;
  color: white;
  position: relative;
  width: 140px;
  margin-right: -140px;
  left: -200px;
  top: -10px;
  height: 35px;
  margin-bottom: -35px;
}