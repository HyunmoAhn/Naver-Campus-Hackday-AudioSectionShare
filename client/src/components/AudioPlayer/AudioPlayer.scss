.AudioPlayer {
  overflow: hidden;
  background-color: #282828;
  border: 1px solid green;
  margin: auto;
  width: 80%;
  text-align: center;
}

.AudioPlayer__play-btn {
  background-color: transparent;
  border: 2px solid #BABABA;
  border-radius: 50%;
  color: white;
  width: 35px;
  height: 35px;
  margin-top: 10px;
  font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
}

.AudioPlayer__section-btn {
  position: relative;
  background-color: transparent;
  border: 0;
  color: white;
  left: -15px;

  .active {
    color: #66D36E;
  }
}

.AudioPlayer__toggle-btn {
  position: relative;
  background-color: transparent;
  border: 0;
  color: white;
  left: 15px;

  .active {
    color: #66D36E;
  }
}

.AudioPlayer__section-loop-cancel {
  background-color: transparent;
  border: 1px solid white;
  color: white;
  position: relative;
  width: 140px;
  margin-right: -140px;
  left: -200px;
  top: -10px;
  height: 35px;
  margin-bottom: -35px;
}