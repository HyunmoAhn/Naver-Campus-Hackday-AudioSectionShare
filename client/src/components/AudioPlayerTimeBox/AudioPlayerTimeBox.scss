.AudioPlayerTimeBox {
  width: 100%;
  overflow: hidden;
}

.AudioPlayerTimeBox__range {
  display: inline-block;
  margin: auto;
  width: 70%;
  text-align: center;
  vertical-align: middle;
}

