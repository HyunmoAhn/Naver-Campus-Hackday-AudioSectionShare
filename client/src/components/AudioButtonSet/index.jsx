import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import TimeScreen from '../TimeScreen';
import './AudioButtonSet.scss';

const propTypes = {
	audio: PropTypes.object,
	endTime: PropTypes.number,
	isSectionLoop: PropTypes.bool,
	isSetSection: PropTypes.bool,
	isShare: PropTypes.bool,
	startTime: PropTypes.number,
	onSectionLoopCancel: PropTypes.func,
	onSetSection: PropTypes.func,
	onShareSetting: PropTypes.func,
	onToggleLoop: PropTypes.func,
	onTogglePlay: PropTypes.func,
};

const defaultProps = {
	audio: {},
	endTime: 0,
	isSectionLoop: false,
	isSetSection: false,
	isShare: false,
	startTime: 0,
	onSectionLoopCancel() {},
	onSetSection() {},
	onShareSetting() {},
	onToggleLoop() {},
	onTogglePlay() {},
};

class AudioButtonSet extends React.Component {
	render() {
		const { endTime, isSectionLoop, isSetSection, isShare, startTime } = this.props;
		const playBtnClassName = cx('fa', {
			'fa-play': this.props.audio.paused,
			'fa-pause': !this.props.audio.paused,
		});
		const loopBtnClassName = cx('fa', 'fa-retweet', {
			'active': this.props.audio.loop,
		});
		const sectionBtnClassName = cx('fa', 'fa-exchange', {
			'active': isSetSection,
		});

		return (
			<div className="AudioButtonSet">
				{isSectionLoop &&
				<button
					className="AudioButtonSet__section-loop-cancel"
					type="button"
					onClick={this.props.onSectionLoopCancel}
				>
					<TimeScreen second={startTime} />
					~
					<TimeScreen second={endTime} />
					<br />
					반복 해제
				</button>
				}
				<button
					className="AudioButtonSet__section-btn"
					disabled={isSectionLoop}
					type="button"
					onClick={this.props.onSetSection}
				>
					<i className={sectionBtnClassName} />
				</button>
				<button
					className="AudioButtonSet__play-btn"
					type="button"
					onClick={this.props.onTogglePlay}
				>
					<i className={playBtnClassName} />
				</button>
				<button
					className="AudioButtonSet__toggle-btn"
					type="button"
					onClick={this.props.onToggleLoop}
				>
					<i className={loopBtnClassName} />
				</button>
				<button
					className="AudioButtonSet__share-btn"
					type="button"
					onClick={this.props.onShareSetting}
				>
					<span className={cx({ 'active': isShare })}>SNS 공유하기</span>
				</button>
			</div>
		);
	}
}

AudioButtonSet.propTypes = propTypes;
AudioButtonSet.defaultProps = defaultProps;

export default AudioButtonSet;