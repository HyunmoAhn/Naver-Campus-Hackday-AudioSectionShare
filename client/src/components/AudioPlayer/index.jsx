import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import AudioPlayerTimeBox from '../AudioPlayerTimeBox';
import AudioPlayerVolumeBox from '../AudioPlayerVolumeBox';
import AudioShare from '../AudioShare';
import TimeScreen from '../TimeScreen';
import './AudioPlayer.scss'

const propTypes = {
	audioInfo: PropTypes.string,
	shareContent: PropTypes.string,
	title: PropTypes.string,
	url: PropTypes.string,
	location: PropTypes.object,
	src: PropTypes.string,
	onFetchAudioInfo: PropTypes.func,
	onShareNaver: PropTypes.func,
	onShareFacebook: PropTypes.func,
};

const defaultProps = {
	audioInfo: '',
	shareContent: '',
	title: '',
	url: '',
	location: {},
	onFetchAudioInfo() {},
	onShareNaver() {},
	onShareFacebook() {},
};

class AudioPlayer extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			currentTime: Number(props.location.query.startTime) || 0,
			isMute: false,
			isLoop: false,
			isPause: true,
			isSetSection: false,
			isSectionLoop: !!props.location.query.startTime && !!props.location.query.endTime,
			isShare: false,
			volume: 0,
			startTime: Number(props.location.query.startTime),
			endTime: Number(props.location.query.endTime),
		};
		this.audio = new Audio();
		this.handleCurrentTimeChange = this.handleCurrentTimeChange.bind(this);
		this.handleToggleMute = this.handleToggleMute.bind(this);
		this.handleTogglePlay = this.handleTogglePlay.bind(this);
		this.handleToggleLoop = this.handleToggleLoop.bind(this);
		this.handleSetSection = this.handleSetSection.bind(this);
		this.handleSectionLoopCancel = this.handleSectionLoopCancel.bind(this);
		this.handleVolumeChange = this.handleVolumeChange.bind(this);
	}

	componentDidMount() {
		const { id } = this.props.location.query;
		this.props.onFetchAudioInfo(id);
	}

	componentWillReceiveProps(nextProps) {
		this.audio = new Audio(nextProps.url);
		this.audio.autoplay = true;
		this.audio.volume = this.state.volume;
		this.audio.onpause = () => {
			if (this.audio.currentTime === this.audio.duration) {
				this.audio.currentTime = 0;
			}
		};
		this.audio.ontimeupdate = () => {
			const { currentTime, startTime, endTime, isSectionLoop } = this.state;

			if (isSectionLoop) {
				if (startTime && currentTime < startTime) {
					this.audio.currentTime = startTime;
				}

				if (endTime && currentTime > endTime) {
					this.audio.currentTime = startTime;
				}
			}

			this.setState({
				currentTime: this.audio.currentTime,
			});
		};
	}

	componentWillUnmount() {
		this.audio.pause();
	}

	handleCurrentTimeChange(e) {
		const duration = this.audio.duration;
		const currentTime = e.target.value / 100 * duration;

		this.audio.currentTime = currentTime;
	}

	handleToggleMute() {
		this.setState({ isMute: !this.audio.muted });
		this.audio.muted = !this.audio.muted;
	}

	handleToggleLoop() {
		this.setState({ isLoop: !this.state.isLoop });
		this.audio.loop = !this.audio.loop;
	}

	handleTogglePlay() {
		this.setState({ isPause: !this.state.isPause });
		if (this.audio.paused) {
			this.audio.play();
		} else {
			this.audio.pause();
		}
	}

	handleSetSection() {
		const { startTime, isSetSection } = this.state;
		const currentTime = Math.round(this.audio.currentTime * 10) / 10;

		this.setState({ isSetSection: !isSetSection });
		if (!isSetSection) {
			this.setState({
				startTime: currentTime,
				endTime: null,
				isSectionLoop: false,
			});
		} else {
			if (currentTime < startTime) {
				this.setState({
					message: '시작시간보다 끝나는 시간이 빠릅니다.',
					startTime: null,
				}, () => {
					setTimeout(() => this.setState({ message: null }), 2000);
				});
				return null;
			}
			this.setState({
				endTime: currentTime,
				isSectionLoop: true,
			});
		}
	}

	handleSectionLoopCancel() {
		this.setState({
			isSectionLoop: false,
			startTime: null,
			endTime: null,
		});
	}

	handleVolumeChange(value) {
		if (this.audio.muted) {
			this.audio.muted = false;
		}
		this.setState({ volume: Number(value) });
		this.audio.volume = value / 100;
	}

	render() {
		const { volume } = this.state;
		const playBtnClassName = cx('fa', {
			'fa-play': this.audio.paused,
			'fa-pause': !this.audio.paused,
		});
		const loopBtnClassName = cx('fa', 'fa-retweet', {
			'active': this.audio.loop,
		});
		const sectionBtnClassName = cx('fa', 'fa-exchange', {
			'active': this.state.isSetSection,
		});

		if (!this.props.url) {
			return <div>
				잠시만 기다려 주세요.
			</div>;
		}

		return (
			<div className="AudioPlayer">
				<div className="AudioPlayer__title">
					{this.props.title}
				</div>
				{this.state.isSectionLoop &&
					<button
						className="AudioPlayer__section-loop-cancel"
					  type="button"
					  onClick={this.handleSectionLoopCancel}
					>
						<TimeScreen second={this.state.startTime} /> ~
						<TimeScreen second={this.state.endTime} />
						<br />
						반복 해제
					</button>
				}
				<button
					className="AudioPlayer__section-btn"
				  type="button"
				  onClick={this.handleSetSection}
				>
					<i className={sectionBtnClassName} />
				</button>
				<button
					className="AudioPlayer__play-btn"
					type="button"
					onClick={this.handleTogglePlay}
				>
					<i className={playBtnClassName} />
				</button>
				<button
					className="AudioPlayer__toggle-btn"
					type="button"
					onClick={this.handleToggleLoop}
				>
					<i className={loopBtnClassName} />
				</button>
				<button
					className="AudioPlayer__share-btn"
				  type="button"
				  onClick={() => {
				  	if (!this.state.isSectionLoop) {
				  		this.setState({ message: '공유 구간을 설정해 주세요.' }, () => {
				  			setTimeout(() => this.setState({ message: '' }), 2000);
						  });
				  		return null;
					  }
					  this.setState({ isShare: !this.state.isShare })
				  }}
				>
					<span className={cx({ 'active': this.state.isShare })}>SNS 공유하기</span>
				</button>
				<AudioPlayerTimeBox
					currentTime={this.audio.currentTime}
				  duration={this.audio.duration}
				  onCurrentTimeChange={this.handleCurrentTimeChange}
				/>
				<AudioPlayerVolumeBox
					volume={this.audio.muted ? 0 : volume}
				  onToggleMute={this.handleToggleMute}
				  onVolumeChange={this.handleVolumeChange}
				/>
				{this.state.message &&
					<div className="AudioPlayer__message">
						{this.state.message}
					</div>
				}
				{this.props.audioInfo &&
					<div className="AudioPlayer__content">
						{this.props.audioInfo}
					</div>
				}
				{this.props.shareContent &&
					<div className="AudioPlayer__share-content">
						{this.props.shareContent}
					</div>
				}
				{this.state.isShare && this.state.isSectionLoop &&
					<AudioShare
						id={this.props.location.query.id}
						startTime={this.state.startTime}
						endTime={this.state.endTime}
						onFacebookShare={this.props.onShareFacebook}
					  onNaverShare={this.props.onShareNaver}
					/>
				}
			</div>
		)
	}
}

AudioPlayer.propTypes = propTypes;
AudioPlayer.defaultProps = defaultProps;

export default AudioPlayer