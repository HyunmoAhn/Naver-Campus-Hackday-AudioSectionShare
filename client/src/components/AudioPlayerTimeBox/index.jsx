import React from 'react';
import PropTypes from 'prop-types';
import TimeScreen from '../TimeScreen';
import './AudioPlayerTimeBox.scss';

const propTypes = {
	currentTime: PropTypes.number,
	duration: PropTypes.number,
	onCurrentTimeChange: PropTypes.func,
};

const defaultProps = {
	currentTime: 0,
	duration: 0,
	onCurrentTimeChange() {},
};

class AudioPlayerTimeBox extends React.Component {
	render() {
		const { currentTime, duration, onCurrentTimeChange } = this.props;
		const relativeCurrentTime = currentTime / duration * 100;

		return (
			<div className="AudioPlayerTimeBox">
				<TimeScreen className="AudioPlayerTimeBox__currentTime" second={currentTime} />
				<div
					className="AudioPlayerTimeBox__range"
				>
					<input
						type="range"
						value={relativeCurrentTime}
						onChange={onCurrentTimeChange}
					/>
				</div>
				<TimeScreen className="AudioPlayerTimeBox__duration" second={duration} />
			</div>
		);
	}
}

AudioPlayerTimeBox.propTypes = propTypes;
AudioPlayerTimeBox.defaultProps = defaultProps;

export default AudioPlayerTimeBox;