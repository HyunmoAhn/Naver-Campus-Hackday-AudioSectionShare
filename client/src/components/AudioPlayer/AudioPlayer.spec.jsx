import React from 'react';
import { shallow, mount } from 'enzyme';

import AudioPlayerTimeBox from 'components/AudioPlayerTimeBox';
import AudioButtonSet from 'components/AudioButtonSet';
import AudioVolumeBox from 'components/AudioPlayerVolumeBox';
import AudioPlayer from './';

describe('<AudioPlayer />', () => {
	const mockUrl = 'http://mock-url.mock';
	it('should match to snapshot when render default', () => {
		const wrapper = shallow(<AudioPlayer />);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match to snapshot when exist props.url', () => {
		const wrapper = shallow(<AudioPlayer url={mockUrl} />);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match to snapshot when exist state.message', () => {
		const wrapper = shallow(<AudioPlayer url={mockUrl} />);
		wrapper.setState({ message: 'mock-message' });

		expect(wrapper).toMatchSnapshot();
	});

	it('should match to snapshot when props.audioInfo', () => {
		const mockAudioInfo = 'mock-audio-info';
		const wrapper = shallow(<AudioPlayer url={mockUrl} audioInfo={mockAudioInfo} />);

		expect(wrapper).toMatchSnapshot();
	});

	it('should match to snapshot when shareContent', () => {
		const mockShareContent = 'mock-share-content';
		const wrapper = shallow(<AudioPlayer url={mockUrl} shareContent={mockShareContent} />);

		expect(wrapper).toMatchSnapshot();
	});

	describe('should test when call componentDidMount', () => {
		it('should call props.onSetTrueSectionLoop when exist startTime and endTime', () => {
			const mockLocation = {
				query: {
					startTime: 10,
					endTime: 20,
				},
			};
			const onSetTrueSectionLoop = jest.fn();
			mount(<AudioPlayer
				location={mockLocation}
				onSetTrueSectionLoop={onSetTrueSectionLoop}
			/>);

			expect(onSetTrueSectionLoop.mock.calls.length).toBe(1);
		});

		it('should call props.onFetchAudioInfo when exist id and isList === false', () => {
			const mockLocation = {
				query: {
					id: 5,
					isList: false,
				},
			};
		const onFetchAudioInfo = jest.fn();
		mount(<AudioPlayer location={mockLocation} onFetchAudioInfo={onFetchAudioInfo} />);

		expect(onFetchAudioInfo.mock.calls.length).toEqual(1);
		expect(onFetchAudioInfo.mock.calls[0]).toEqual([mockLocation.query.id]);
		});
	});

	describe('should test when call componentWillReceiveProps', () => {
		it('should call correct in onPause when this.audio.currentTime === this.audio.duration', () => {
			const wrapper = mount(<AudioPlayer />);
			let audio = wrapper.instance().audio;

			wrapper.setProps({ url: 'mock-url' });
			audio = wrapper.instance().audio;
			audio.currentTime = 100;
			audio.onpause();
			expect(audio.currentTime).toBe(100);
			audio.currentTime = audio.duration;
			audio.onpause();

			expect(audio.currentTime).toBe(0);
		});

		describe('should call correct in ontimeupdate', () => {
			const wrapper = mount(<AudioPlayer />);
			let audio = wrapper.instance().audio;
			wrapper.setProps({ url: 'mock-url' });
			audio = wrapper.instance().audio;

			describe('when if props.isSectionLoop === true', () => {
				wrapper.setProps({ isSectionLoop: true });
				it('when if audio.currentTime < startTime', () => {
					wrapper.instance().audio.currentTime = 10;
					wrapper.setState({ startTime: 30 });
					audio.ontimeupdate();
					expect(wrapper.instance().audio.currentTime).toBe(30);
					expect(wrapper.instance().state.currentTime).toBe(30);
				});

				it('when if audio.currentTime >= startTime', () => {
					wrapper.instance().audio.currentTime = 100;
					wrapper.setState({ startTime: 30, endTime: 150 });
					audio.ontimeupdate();
					expect(wrapper.instance().audio.currentTime).toBe(100);
					expect(wrapper.instance().state.currentTime).toBe(100);
				});
			});

			it('when if props.isSectionLoop === true and audio.currentTIme > endTime', () => {
				wrapper.setProps({ isSectionLoop: true });
				audio.currentTime = 100;
				wrapper.setState({ startTime: 30, endTime: 50 });
				audio.ontimeupdate();
				expect(wrapper.instance().audio.currentTime).toBe(30);
				expect(wrapper.instance().state.currentTime).toBe(30);
			});

			it('when if props.isSectionLoop !== true', () => {
				wrapper.setProps({ isSectionLoop: false });
				wrapper.instance().audio.currentTime = 150;
				audio.ontimeupdate();
				expect(wrapper.instance().state.currentTime).toBe(150);
			});
		});

		it('should call handleCurrentTimeChange when call AudioPlayerTimeBox.onCurrentTimeChange', () => {
			const handleCurrentTimeChange = jest.spyOn(AudioPlayer.prototype, 'handleCurrentTimeChange');
			const wrapper = shallow(<AudioPlayer url={mockUrl} />);
			const audioPlayerTimeBox = wrapper.find(AudioPlayerTimeBox);
			const duration = wrapper.instance().audio.duration;
			const mockEvent = {
				target: {
					value: 100,
				},
			};
			const currentTime = mockEvent.target.value / 100 * duration;
			audioPlayerTimeBox.props().onCurrentTimeChange(mockEvent);

			expect(handleCurrentTimeChange.mock.calls.length).toBe(1);
			expect(handleCurrentTimeChange.mock.calls[0]).toEqual([mockEvent]);
			expect(wrapper.instance().audio.currentTime).toBe(currentTime);
		});

		describe('test AudioButtonSet Component Event', () => {
			const onSetFalseSectionLoop = jest.fn();
			const onSetTrueSectionLoop = jest.fn();
			const handleSectionLoopCancel = jest.spyOn(AudioPlayer.prototype, 'handleSectionLoopCancel');
			const handleSetSection = jest.spyOn(AudioPlayer.prototype, 'handleSetSection');
			const handleShareSetting = jest.spyOn(AudioPlayer.prototype, 'handleShareSetting');
			const handleToggleLoop = jest.spyOn(AudioPlayer.prototype, 'handleToggleLoop');
			const handleTogglePlay = jest.spyOn(AudioPlayer.prototype, 'handleTogglePlay');
			const wrapper = shallow(<AudioPlayer
				url={mockUrl}
			  onSetFalseSectionLoop={onSetFalseSectionLoop}
				onSetTrueSectionLoop={onSetTrueSectionLoop}
			/>);
			const audioButtonSet = wrapper.find(AudioButtonSet);

			it('should call handleSectionLoopCancel when call onSectionLoopCancel', () => {
				audioButtonSet.props().onSectionLoopCancel();

				expect(handleSectionLoopCancel.mock.calls.length).toBe(1);
				expect(wrapper.state().startTime).toBe(null);
				expect(wrapper.state().endTime).toBe(null);
				expect(onSetFalseSectionLoop.mock.calls.length).toBe(1);
				onSetFalseSectionLoop.mockClear();
			});

			describe('should call handleSetSection when call onSetSection', () => {
				wrapper.instance().audio.currentTime = 10.555;
				const currentTime = Math.round(wrapper.instance().audio.currentTime * 10) / 10;

				it('when isSetSection === false', () => {
					wrapper.setState({ isSetSection: false });
					audioButtonSet.props().onSetSection();

					expect(handleSetSection.mock.calls.length).toBe(1);
					expect(wrapper.state().isSetSection).toBe(true);
					expect(wrapper.state().startTime).toBe(currentTime);
					expect(wrapper.state().endTime).toBe(null);
					expect(onSetFalseSectionLoop.mock.calls.length).toBe(1);
					onSetFalseSectionLoop.mockClear();
					handleSetSection.mockClear();
				});

				it('when isSection === true', () => {
					wrapper.setState({ isSetSection: true });
					audioButtonSet.props().onSetSection();

					expect(handleSetSection.mock.calls.length).toBe(1);
					expect(wrapper.state().isSetSection).toBe(false);
					expect(wrapper.state().endTime).toBe(currentTime);
					expect(onSetTrueSectionLoop.mock.calls.length).toBe(1);

					handleSetSection.mockClear();
					onSetTrueSectionLoop.mockClear();
				});

				it('when isSection === true and currentTime < startTime', () => {
					wrapper.setState({
						isSetSection: true,
						startTime: 15,
					});
					audioButtonSet.props().onSetSection();

					expect(handleSetSection.mock.calls.length).toBe(1);
					expect(wrapper.state().message).toBe('시작시간보다 끝나는 시간이 빠릅니다.');
					expect(wrapper.state().startTime).toBe(null);
					handleSetSection.mockClear();
				});
			});

			it('should call handleShareSetting when call onShareSetting', () => {
				wrapper.setProps({
					location: {
						query: {
							id: 'mock-id'
						},
					},
					isSectionLoop: true,
				});
				wrapper.setState({
					startTime: 10,
					isShare: true,
				});
				audioButtonSet.props().onShareSetting();
				expect(wrapper.state().isShare).toBe(false);

				wrapper.setProps({
					isSectionLoop: false,
				});
				audioButtonSet.props().onShareSetting();
				expect(wrapper.state().message).toBe('공유 구간을 설정해 주세요.');
			});

			it('should call handleToggleLoop when call onToggleLoop', () => {
				wrapper.instance().audio.loop = false;
				audioButtonSet.props().onToggleLoop();

				expect(handleToggleLoop.mock.calls.length).toBe(1);
				expect(wrapper.instance().audio.loop).toBe(true);
				handleToggleLoop.mockClear();
			});

			it.skip('should call handleTogglePlay when call onTogglePlay', () => {});

			describe('simulate AudioPlayerVolumeBox.props', () => {
				const handleToggleMute = jest.spyOn(AudioPlayer.prototype, 'handleToggleMute');
				const handleVolumeChange = jest.spyOn(AudioPlayer.prototype, 'handleVolumeChange');
				const wrapper = shallow(<AudioPlayer url={mockUrl} />);
				const volumeBox = wrapper.find(AudioVolumeBox);

				it('should call handleToggleMute when simulate onToggleMute', () => {
					volumeBox.props().onToggleMute();

					expect(handleToggleMute.mock.calls.length).toBe(1);
					expect(wrapper.instance().audio.muted).toBe(true);
				});

				it('should call handleVolumeChange when simulate onVolumeChange', () => {
					const mockVolume = 50.534;
					volumeBox.props().onVolumeChange(mockVolume);

					expect(wrapper.instance().audio.muted).toBe(false);
					expect(wrapper.state().volume).toBe(Number(mockVolume));
					expect(wrapper.instance().audio.volume).toBe(mockVolume / 100);
				});
			});
		});
	})
});