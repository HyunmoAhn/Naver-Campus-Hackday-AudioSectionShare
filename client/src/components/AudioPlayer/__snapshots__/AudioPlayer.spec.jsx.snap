// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AudioPlayer /> should match to snapshot when exist props.url 1`] = `
<div
  className="AudioPlayer"
>
  <div
    className="AudioPlayer__title"
  />
  <AudioButtonSet
    audio={
      <audio
        preload="auto"
      />
    }
    endTime={NaN}
    isSectionLoop={false}
    isSetSection={false}
    isShare={false}
    onSectionLoopCancel={[Function]}
    onSetSection={[Function]}
    onShareSetting={[Function]}
    onToggleLoop={[Function]}
    onTogglePlay={[Function]}
    startTime={NaN}
  />
  <AudioPlayerTimeBox
    currentTime={0}
    duration={0}
    onCurrentTimeChange={[Function]}
  />
  <AudioPlayerVolumeBox
    onToggleMute={[Function]}
    onVolumeChange={[Function]}
    volume={50}
  />
</div>
`;

exports[`<AudioPlayer /> should match to snapshot when exist state.message 1`] = `
<div
  className="AudioPlayer"
>
  <div
    className="AudioPlayer__title"
  />
  <AudioButtonSet
    audio={
      <audio
        preload="auto"
      />
    }
    endTime={NaN}
    isSectionLoop={false}
    isSetSection={false}
    isShare={false}
    onSectionLoopCancel={[Function]}
    onSetSection={[Function]}
    onShareSetting={[Function]}
    onToggleLoop={[Function]}
    onTogglePlay={[Function]}
    startTime={NaN}
  />
  <AudioPlayerTimeBox
    currentTime={0}
    duration={0}
    onCurrentTimeChange={[Function]}
  />
  <AudioPlayerVolumeBox
    onToggleMute={[Function]}
    onVolumeChange={[Function]}
    volume={50}
  />
  <div
    className="AudioPlayer__message"
  >
    mock-message
  </div>
</div>
`;

exports[`<AudioPlayer /> should match to snapshot when props.audioInfo 1`] = `
<div
  className="AudioPlayer"
>
  <div
    className="AudioPlayer__title"
  />
  <AudioButtonSet
    audio={
      <audio
        preload="auto"
      />
    }
    endTime={NaN}
    isSectionLoop={false}
    isSetSection={false}
    isShare={false}
    onSectionLoopCancel={[Function]}
    onSetSection={[Function]}
    onShareSetting={[Function]}
    onToggleLoop={[Function]}
    onTogglePlay={[Function]}
    startTime={NaN}
  />
  <AudioPlayerTimeBox
    currentTime={0}
    duration={0}
    onCurrentTimeChange={[Function]}
  />
  <AudioPlayerVolumeBox
    onToggleMute={[Function]}
    onVolumeChange={[Function]}
    volume={50}
  />
  <div
    className="AudioPlayer__content"
  >
    mock-audio-info
  </div>
</div>
`;

exports[`<AudioPlayer /> should match to snapshot when render default 1`] = `
<div>
  잠시만 기다려 주세요.
</div>
`;

exports[`<AudioPlayer /> should match to snapshot when shareContent 1`] = `
<div
  className="AudioPlayer"
>
  <div
    className="AudioPlayer__title"
  />
  <AudioButtonSet
    audio={
      <audio
        preload="auto"
      />
    }
    endTime={NaN}
    isSectionLoop={false}
    isSetSection={false}
    isShare={false}
    onSectionLoopCancel={[Function]}
    onSetSection={[Function]}
    onShareSetting={[Function]}
    onToggleLoop={[Function]}
    onTogglePlay={[Function]}
    startTime={NaN}
  />
  <AudioPlayerTimeBox
    currentTime={0}
    duration={0}
    onCurrentTimeChange={[Function]}
  />
  <AudioPlayerVolumeBox
    onToggleMute={[Function]}
    onVolumeChange={[Function]}
    volume={50}
  />
  <div
    className="AudioPlayer__share-content"
  >
    mock-share-content
  </div>
</div>
`;
